common/persistence/cassandra/nexus_service_store.go

```diff
@@ -61,30 +104,282 @@ func (s *NexusServiceStore) Close() {
+
+	if request.Service.Version == 0 {
+		batch.Query(templateCreateIncomingServiceQuery,
+			constServicesPartition,
+			rowTypeIncomingNexusService,
+			request.Service.ServiceID,
+			request.Service.Data.Data,
+			request.Service.Data.EncodingType.String(),
+			1,
+		)
+	} else {
+		batch.Query(templateUpdateIncomingServiceQuery,
+			request.Service.Data.Data,
+			request.Service.Data.EncodingType.String(),
+			request.Service.Version+1,
+			constServicesPartition,
+			rowTypeIncomingNexusService,
+			request.Service.ServiceID,
+			request.Service.Version,
+		)
+	}
+
+	if request.LastKnownTableVersion == 0 {
+		batch.Query(templateCreateTableVersion,
+			constServicesPartition,
+			rowTypePartitionStatus,
+			constTableVersionServiceID,
+			1)
+	} else {
+		batch.Query(TemplateUpdateTableVersion,
+			request.LastKnownTableVersion+1,
+			constServicesPartition,
+			rowTypePartitionStatus,
+			constTableVersionServiceID,
+			request.LastKnownTableVersion)
+	}
+
+	previousPartitionStatus := make(map[string]interface{})
+	applied, iter, err := s.session.MapExecuteBatchCAS(batch, previousPartitionStatus)
+
+	if err != nil {
+		return gocql.ConvertError("CreateOrUpdateNexusIncomingService", err)
+	}
+
+	previousService := make(map[string]interface{})
+	iter.MapScan(previousService)
+
+	err = iter.Close()
+	if err != nil {
+		return gocql.ConvertError("CreateOrUpdateNexusIncomingService", err)
+	}
+
+	if !applied {
+		currentTableVersion, err := getTypedFieldFromRow[int64]("version", previousPartitionStatus)
+		if err != nil {
+			return fmt.Errorf("error retrieving current table version: %w", err)
+		}
+		if currentTableVersion != request.LastKnownTableVersion {
+			return fmt.Errorf("%w. Provided table version: %v Current table version: %v",
+				ErrTableVersionConflict,
+				request.LastKnownTableVersion,
+				currentTableVersion)
+		}
+
+		currentServiceVersion, err := getTypedFieldFromRow[int64]("version", previousService)
+		if err != nil {
+			return fmt.Errorf("error retrieving current service version: %w", err)
+		}
+		if currentServiceVersion != request.Service.Version {
+			return fmt.Errorf("%w. Provided service version: %v Current service version: %v",
+				ErrNexusIncomingServiceVersionConflict,
+				request.Service.Version,
+				currentServiceVersion)
+		}
+
+		// This should never happen. This means the request had the correct versions and gocql did not
+		// return an error but for some reason the update was not applied.
+		return serviceerror.NewInternal("CreateOrUpdateNexusIncomingService failed.")
+	}
+
+	return nil
 }
 
 func (s *NexusServiceStore) ListNexusIncomingServices(
 	ctx context.Context,
 	request *p.InternalListNexusIncomingServicesRequest,
 ) (*p.InternalListNexusIncomingServicesResponse, error) {
-	return nil, serviceerror.NewUnimplemented("Cassandra ListNexusIncomingServices is not implemented")
+	if request.PageSize <= 0 {
+		return nil, ErrNonPositiveListNexusIncomingServicesPageSize
```

Comment:
It's okay if you put this validation here but there should also be validation in the frontend that should return a bad request error (for the next PR).